File: medfuncBackprojection.cpp
#include<vector>

std::vector<std::vector<double>> medfuncBackprojection(int theta_num, int N, std::vector<std::vector<double>> R, double delta) {
	std::vector<std::vector<double>> rec(N, std::vector<double>(N, 0));
	for (int i = 0; i < theta_num; i++) {
		std::vector<std::vector<double>> pm(N, std::vector<double>(1, 0));
		for (int j = 0; j < N; j++) {
			pm[j][0] = R[j][i];
		}
		double Cm = (N / 2) * (1 - cos(i * delta) - sin(i * delta));
		for (int j = 0; j < N; j++) {
			for (int k = 0; k < N; k++) {
				double Xrm = Cm + k * cos(i * delta) + j * sin(i * delta);
				int n = floor(Xrm);
				double t = Xrm - n;
				if (n < 0) n = 0;
				if (n > N - 2)n = N - 2;
				double p = (1 - t) * pm[n][0] + t * pm[n + 1][0];
				rec[k][j] = rec[k][j] + p;
			}
		}
	}
	return rec;
}

File: medfuncBackprojection.m
function rec=medfuncBackprojection(theta_num,N,R1,delta)
rec=zeros(N);
for m=1:theta_num
    pm=R1(:,m);
    Cm=(N/2)*(1-cos((m-1)*delta)-sin((m-1)*delta));
    for k1=1:N
        for k2=1:N
            Xrm=Cm+(k2-1)*cos((m-1)*delta)+(k1-1)*sin((m-1)*delta);
            n=floor(Xrm);
            t=Xrm-floor(Xrm);
            n=max(1,n);n=min(n,N-1);
            p=(1-t)*pm(n)+t*pm(n+1);
            rec(N+1-k1,k2)=rec(N+1-k1,k2)+p;
        end
    end
end
end